### PR TITLE
Add an efficient text stream write function

### PR DESCRIPTION
**Pull request by lhecker**
This adds PR adds a couple foundational functions and classes to make
our TextBuffer more performant and allow us to improve our Unicode
correctness in the future, by getting rid of our dependence on
`OutputCellIterator`. In the future we can then replace the simple
UTF-16 code point iterator with a proper grapheme cluster iterator.

While my focus is technically on Unicode correctness, the ~4x VT
throughput increase in OpenConsole is pretty nice too.

This PR adds:
* A new, simpler ROW iterator (unused in this PR)
* Cursor movement functions (`NavigateToPrevious`, `NavigateToNext`)
  They're based on functions that align the cursor to the start/end
  of the _current_ cell, so such functions can be added as well.
* `ReplaceText` to write a raw string of text with the possibility to
  specify a right margin.
* `CopyRangeFrom` will allow us to make reflow much faster, as it's able
  to bulk-copy already measured strings without re-measuring them.

Related to #8000

## Validation Steps Performed
* enwik8.txt, zhwik8.txt, emoji-test.txt, all work with proper
  wide glyph reflow at the end of a row ✅
* This produces "a 咪" where only "a" has a white background:
  ```sh
  printf '\e7こん\e8\x1b[107ma\x1b[m\n'
  ```
* This produces "abん":
  ```sh
  stdbuf -o0 printf '\x1b7こん\x1b8a'; printf 'b\n'
  ```
* This produces "xy" at the end of the line:
  ```sh
  stdbuf -o0 printf '\e[999C\bこ\bx'; printf 'y\n'
  ```
* This produces red whitespace followed by "こ " in the default
  background color at the end of the line, and "ん" on the next line:
  ```sh
  printf '\e[41m\e[K\e[m\e[999C\e[2Dこん\n'
  ```

### CODE REVIEW

**Comment by zadjii-msft** [comment]
me reading this pr body rn

image

File: src/terminal/adapter/adaptDispatch.cpp [hunk 1]
```diff
@@ -101,45 +101,31 @@ void AdaptDispatch::_WriteToBuffer(const std::wstring_view string)
...
-            cursorPosition.x = lineWidth;
-            // But if wrapping is disabled, we also need to move to the next string
-            // position, otherwise we'll be stuck in this loop forever.
-            if (!wrapAtEOL)
-            {
-                stringPosition++;
-            }
-        }
-        else
+        const auto newPosX = textBuffer.Write(cursorPosition.y, cursorPosition.x, til::CoordTypeMax, wrapAtEOL, attributes, stringIterator);
```
Comment (j4james) [inline src/terminal/adapter/adaptDispatch.cpp:120]: Why isn't the `lineWidth` being passed to this any more? It looks to me like it's not going to wrap correctly when you're on a double-width line.

Comment (lhecker) [reply src/terminal/adapter/adaptDispatch.cpp:120]: Ah good catch! I'll add a comment to `lineWidth` to point out its importance.

File: tools/ConsoleTypes.natvis [hunk 2]
```diff
@@ -38,31 +38,12 @@
...
-        <DisplayString>{{ wrap={_wrapForced} padded={_doubleBytePadded} }}</DisplayString>
-        <Expand>
-            <ExpandedItem>_data</ExpandedItem>
-        </Expand>
-    </Type>
-
     <Type Name="ROW">
-        <DisplayString>{{ id={_id} width={_rowWidth} }}</DisplayString>
+        <DisplayString>{_chars.data(),[_charOffsets[_columnCount]]}</DisplayString>
+		<StringView>_chars.data(),[_charOffsets[_columnCount]]</StringView>
```
Comment (DHowett) [inline tools/ConsoleTypes.natvis:43]: The spacing in this file seems to have gone bad!

File: src/inc/test/CommonState.hpp [hunk 3]
```diff
@@ -256,49 +256,20 @@ class CommonState
...
-        };
-
-        applyTestString(pRow, testStrings);
+        uint16_t column = 0;
+        for (const auto& ch : std::wstring_view{ L"AB\u304bC\u304dDE      " })
+        {
+            const uint16_t width = ch >= 0x80 ? 2 : 1;
+            pRow->ReplaceCharacters(column, width, { &ch, 1 });
+            column += width;
+        }
```
Comment (lhecker) [inline src/inc/test/CommonState.hpp:272]: I forgot why I made this change... But it's much shorter now, so there's that at least.

File: src/buffer/out/Row.cpp [hunk 4]
```diff
@@ -65,6 +68,65 @@ constexpr OutIt copy_n_small(InIt first, Diff count, OutIt dest)
     return dest;
 }
 
+RowTextIterator::RowTextIterator(std::span<const wchar_t> chars, std::span<const uint16_t> charOffsets, uint16_t offset) noexcept :
```
Comment (lhecker) [inline src/buffer/out/Row.cpp:71]: `RowTextIterator` isn't used anywhere yet, but I did use it during development as a debug aid. I expect this struct to be used in the future for things like `CHAR_INFO` reads.

**Comment by lhecker** [comment]
I began writing a new implementation for `TextBuffer::Reflow` based on first principles over at [dev/lhecker/8000-stream-writing-2](https://github.com/microsoft/terminal/tree/dev/lhecker/8000-stream-writing-2)
Even at roughly 400x200 large viewports I can now resize OpenConsole at >120 FPS, whereas it previously ran at about 10 FPS for me. The difference is quite satisfying.

Unfortunately it still pegs an entire CPU core, but it'll be equally fun and satisfying to fix that: We can still...
* implement lazy initialization for `ROW` so that we don't have to first fill all those `chars` with whitespace and `charOffsets` with 0, 1, 2, 3, 4, ...
* try to replace `GetLastNonSpaceCharacter` with our existing virtual bottom logic and/or avoid running it twice on OpenConsole's side by better integrating `SCREEN_INFORMATION::ResizeWithReflow`
* vectorize `ROW::WriteHelper::CopyRangeFrom`, because it is the innermost loop during resize and vectorization is quite trivial there (I added the code already, it's commented out though)

Unfortunately, for the life of me, I can't figure out how to properly implement cursor wrapping. From what I learned about our VT code I would've assumed that I need to use `Cursor::DelayEOLWrap()`, but the old code didn't use it at all. So I'm left wondering now how the old code "unwrapped" cursors when widening the viewport. 😅 I mean I do get how the old code works, but I'm failing to understand how it works arithmetically, so that I can rewrite it.

File: src/terminal/adapter/adaptDispatch.cpp [hunk 5]
```diff
@@ -101,60 +101,54 @@ void AdaptDispatch::_WriteToBuffer(const std::wstring_view string)
...
-            cursor.SetPosition(cursorPosition);
             if (wrapAtEOL)
             {
                 cursor.DelayEOLWrap();
             }
-        }
-        else
-        {
-            cursor.SetPosition(cursorPosition);
+            else if (failedToWrite)
```
Comment (DHowett) [inline src/terminal/adapter/adaptDispatch.cpp:144]: Are we certain there's no other instances in which this could hit?

File: src/buffer/out/Row.cpp [hunk 6]
```diff
@@ -499,6 +739,11 @@ void ROW::_resizeChars(uint16_t colExtEnd, uint16_t chExtBeg, uint16_t chExtEnd,
     }
 }
 
+til::small_rle<TextAttribute, uint16_t, 1>& ROW::Attributes() noexcept
```
Comment (DHowett) [inline src/buffer/out/Row.cpp:742]: used?

**Review by DHowett — APPROVED** [review]
I think I understand this, and I need a second adult to think they also understand it.

File: src/terminal/adapter/adaptDispatch.cpp (hunk 5 above)
Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:144]: I would also expect this to happen when we're writing a wide glyph in the last column and autowrap mode _is_ enabled. Because the `TextBuffer::Write` can't handle the wrapping (or if it is, it probably shouldn't be), so it has to fail and let us repeat the loop with the delayed wrap flag set.

When I have a chance I'll try and give this branch a test, but I'm almost certain there's something wrong here.

Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:144]: Disregard that - this case is fine. `TextBuffer::Write` still advances the cursor position, so `failedToWrite` is true, but `isWrapping` should also be true, because `newPosX` will have reached the `lineWidth`.

Comment (lhecker) [reply src/terminal/adapter/adaptDispatch.cpp:144]: `TextBuffer::Write` always returns:
* `colBeg` if the parameters are invalid (= it return the column you gave as the function argument)
* `colExtBeg` otherwise, which is the column past the last character that was written. This includes any padding space that is introduced because a wide character failed to insert!

So in case that we write the last column, and assuming the parameters are valid, this return value will _always_ be equal to `lineWidth`. This is very important, because that's how we detect line wrapping (`isWrapping` is true if `TextBuffer::Write` returns `lineWidth`).

In short, `failedToWrite` can only ever be true if `isWrapping` is true.
...unless I'm mistaken of course.

I'd be extremely happy if you could take a look at this code btw @j4james, if you got the time. Of course this is entirely optional, but the last thing I'd ever want is for this PR to stand in the way of any future plans you have and so if you have any feedback, no matter how critical it is, I'd love to hear it.

**Comment by j4james** [comment]
I've just tried out this branch, and there is one change in behavior which I think is potentially bad. If you have a wide glyph output with a black background, and you overwrite the start of it with a narrow glyph using a white background, both cells that were previously occupied by the wide glyph are changed from black to white. In the previous implementation, the second cell would remain black, which I think was the correct behavior.

An example of where this can be a problem is when you have a popup window with a white background on top of some wide glyphs using a black background. Whenever the border of the window intersects with a wide glyph it'll end up filling more cells that intended, giving the window a broken effect. See below:

image

**Comment by lhecker** [comment]
Hmm unless I forgot something I don't think we need the ability to write text _without_ modifying attributes right now, do we? At least not in any fast path I mean (in a hypothetical, seldomly used slow path we can just make a backup of the row and restore the attributes from the backup).

Because then I could just move writing attributes into the `ROW::Write` function which gives it access to the `colEnd` variable (the last written column, excluding padding space).

**Comment by j4james** [comment]
I've just noticed another problem (possibly related to the issue above). When you write a narrow glyph over the start of a wide glyph, the cursor position is moved up by two. So if you try to write a line of English text over some Japanese text, you end up with everything double-spaced!

File: src/terminal/adapter/adaptDispatch.cpp (hunk 5 above)
Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:144]: I'll try and take a look at the code later this weekend, but the `TextBuffer` area is a bit out of my comfort zone, so I probably wouldn't be able to provide a very useful review. That said, I have just rebased all my pending work on top of this branch, and I didn't have any problems with that. So I can at least confirm that this isn't messing with any of my current plans.

**Comment by lhecker** [comment]
I've tested the first issue with this:
```sh
printf '\x1b7猫咪\x1b8\x1b[107ma\x1b[m\n'
```
And the second issue with this:
```sh
stdbuf -o0 printf '\x1b7猫咪\x1b8a'; printf 'b\n'
```

As far as I can tell the latest commit fixes both issues. Thank you so much for finding these!

File: src/inc/LibraryIncludes.h [hunk 7]
```diff
@@ -17,6 +17,8 @@
 
 // Block minwindef.h min/max macros to prevent <algorithm> conflict
 #define NOMINMAX
+// Exclude rarely-used stuff from Windows headers
+#define WIN32_LEAN_AND_MEAN
```
Comment (lhecker) [inline src/inc/LibraryIncludes.h:21]: This prevents the inclusion of `commdlg.h` in the `BufferOut` project, which causes a conflict with `ROW::ReplaceText` (`ReplaceText` is a macro). I wish we could somehow disable all these `TCHAR` macros in the `/um/` headers.

**Comment by lhecker** [comment]
(I'll finish fixing the unit tests Monday and add some tests for the 2 issues.)

> I'm not positive, but I thought that's what `WriteConsoleOutputCharacter` does.

For now I kept the existing approach of having separate text- and attribute-writing methods on `ROW`. `TextBuffer` can then call them as it sees fit. So in case of `WriteConsoleOutputCharacter` I imagine we can just add another `TextBuffer` method which only calls `ReplaceText` and handles line wrapping in a loop. There's also the option to make backups of the row's attributes and restore the backup afterwards. I'm personally not too concerned about the overhead there, because the overhead of the cell-wise iterator-based APIs we have right now are way beyond anything that involves making a few memory copies. Basically, I'm confident that I'll be able to replicate anything that `OutputCellIterator` can do now in the future as well, no matter what.

But I'm really not sure what the best approach is for the new "batch writing" APIs I'm adding in this PR. I don't like the design this PR uses at all, but I'm also failing to come up with anything better. There's of course the option to keep something like `OutputCellIterator` and continue to only have a single `ROW:Write` function, which then internally has a switch/case to specialize on all the different kinds a `OutputCellIterator` can be (text writing, attribute writing, text and attribute writing, filling with characters, etc.).

While initially coding this, I thought that splitting it up into distinct purposes (making text and attribute writing separate) would allow us to better compose them back together to implement our various APIs, make it easier to test the now smaller behavior, implement extended grapheme clusters (which requires access to full strings and not just iterators). Now, I'm not too sure anymore about the design aspect.

Although I am already quite happy with how much it simplified `CommandListPopup.cpp` despite these shortcomings: https://github.com/microsoft/terminal/commit/ae0f682965504410ea4f578499aa728dc9e11ba1#diff-0e5f3869db09d92c3424d43a535efe53edfff1a1a4b8fb95081f39839a172179

File: src/buffer/out/Row.cpp (hunk 6 above)
Comment (lhecker) [reply src/buffer/out/Row.cpp:753]: I'm currently using it in a follow-up branch. I can remove it for this PR, but I also feel like it doesn't hurt keeping it in.

File: src/buffer/out/textBuffer.cpp [hunk 8]
```diff
@@ -376,6 +376,24 @@ bool TextBuffer::_PrepareForDoubleByteSequence(const DbcsAttribute dbcsAttribute
...
+    chars = til::utf16_pop(chars);
+}
+
+void TextBuffer::Write(til::CoordType row, bool wrapAtEOL, const TextAttribute& attributes, RowWriteState& state)
+{
+    auto& r = GetRowByOffset(row);
+
+    r.ReplaceText(state);
+    r.ReplaceAttributes(state.columnBegin, state.columnEnd, attributes);
+    r.SetWrapForced(wrapAtEOL && state.columnEndDirty >= r.size());
```
Comment (j4james) [inline src/buffer/out/textBuffer.cpp:392]: This is a bit of an edge case, but if you've got one column free at the end of the line, and you're trying to write out a wide glyph, you won't replace any attributes because `columnEnd` will equal `columnBegin`. But this is one case when I think that `ReplaceAttributes` call should possibly be using `columnEndDirty` (not sure of the exact logic, but something along those lines). Because even though you haven't stored anything in that last column, it has effectively been overwritten and should really take the active attribute color.

For example, in the test case below I've got a couple of wide glyphs written with a default background color, that wrap over the end of a line that was originally red. It feels wrong that you appear to end up with a red space between the characters.

    printf "\e[41m\e[K\e[m\e[999C\e[2Dこん\n"

Although if this is a pain to get right, I don't think it's a big deal.

File: src/buffer/out/textBuffer.cpp [hunk 9]
```diff
@@ -376,6 +376,24 @@ bool TextBuffer::_PrepareForDoubleByteSequence(const DbcsAttribute dbcsAttribute
...
+
+    TriggerRedraw(Viewport::FromExclusive({ state.columnBegin, row, state.columnEnd, row + 1 }));
```
Comment (j4james) [inline src/buffer/out/textBuffer.cpp:394]: This may not matter if we're doing full row redraws anyway, but shouldn't this be  using `columnEndDirty`?

**Review by j4james — COMMENTED** [review]
I've read through everything now, but the `ROW` internals were a bit over my head so I'm afraid I just skimmed most of that.

<s>Other than the comments I made in the code, there was one other regression I noticed in testing, but I couldn't see why it was failing. It's quite possible it's not this PR that's to blame, but I'm just making a note of it here so it doesn't get lost.

    printf "\e[15832;16;27;22;54\$x"

That should fill a block with a width of 14 glyphs, taking up 28 columns, but it's now writing out 28 glyphs, taking up 56 columns.</s>

File: src/buffer/out/Row.cpp [hunk 10]
```diff
@@ -65,6 +68,65 @@ constexpr OutIt copy_n_small(InIt first, Diff count, OutIt dest)
...
+DbcsAttribute RowTextIterator::DbcsAttr() const noexcept
+{
+    return Cols() == 2 ? DbcsAttribute::Leading : DbcsAttribute::Single;
+}
+
+// Safety: col must be [0, _columnCount].
+uint16_t RowTextIterator::_uncheckedCharOffset(size_t col) const noexcept
+{
+    assert(col <= _charOffsets.size());
+    return til::at(_charOffsets, col) & CharOffsetsMask;
```
[the remark this answers is not in the document]
Comment (j4james) [reply src/buffer/out/Row.cpp:120]: Assuming I have understood this correctly, and this assert is off by one, then I think the `operator++` implementation may be off by one as well.

**Comment by j4james** [comment]
> That should fill a block with a width of 14 glyphs, taking up 28 columns, but it's now writing out 28 glyphs, taking up 56 columns.

You can ignore this. It seems I checked out your branch when there was a "wip" commit that was the source of this issue.

File: src/buffer/out/textBuffer.cpp (hunk 8 above)
Comment (j4james) [reply src/buffer/out/textBuffer.cpp:392]: I also wanted to note that the `SetWrapForced` call is probably wrong. Writing to the last column shouldn't treat the row as wrapped. That should only happen once you output an additional character that forces the cursor to move to the next line. But this is an existing bug (#3088), so I wouldn't necessarily expect it to be addressed in this PR.

File: src/buffer/out/textBuffer.cpp (hunk 9 above)
Comment (lhecker) [reply src/buffer/out/textBuffer.cpp:394]: Oh that's interesting... It should also be `state.columnBeginDirty`! But that code behaves identical to the current state (before this PR). Unless I'm mistaken, I think this is a regression due to #13626, which started to properly clear leading/trailing half of wide glyphs and this `TriggerRedraw` should've been fixed in that PR.

File: src/buffer/out/textBuffer.cpp (hunk 8 above)
Comment (lhecker) [reply src/buffer/out/textBuffer.cpp]: BTW I believe this line has another bug: For rows with a double-width line-rendition, we need to check for `state.columnEndDirty >= r.size() / 2`. In my latest commit I've renamed the function to `WriteLine`. It now also automatically gets the line width from the row, so we don't have to get and update it in `AdaptDispatch::_WriteToBuffer` anymore.

I believe the `SetWrapForced` bug you mentioned can be fixed by doing this:
```cpp
r.SetWrapForced(wrapAtEOL && state.columnEnd >= state.columnLimit && !state.text.empty());
```

In other words: It only wraps if the new cursor position (`columnEnd`) is past the line width (`columnLimit`) but there's more text to come (`!text.empty()).

However I found that this breaks the current implementation of `TextBuffer::Reflow`, which has this condition for line breaks: https://github.com/microsoft/terminal/blob/b9248fa903eaa8d187c72bf08f8aa17b25f9d6bb/src/buffer/out/textBuffer.cpp#L2486-L2491

My latest commit removes the `iRight < cOldColsTotal` part of the condition, which then fixes the bug as described in #3088 for both widening and narrowing the viewport. I'm not sure if removing that part of the condition is safe however, because even with the code comment above it, I can't quite figure out why the check exists in the first place. Isn't `row.WasWrapForced()` indication enough that a row had an explicit newline? [^1] I might drop this part of the PR before it merges if I continue  feeling that unconfident about it. 😅

[^1]: My understanding so far is: It always wraps as if a `\r\n` was there, _unless_ `WasWrapForced()` is `true`.

Comment (j4james) [reply src/buffer/out/textBuffer.cpp]: > In other words: It only wraps if the new cursor position (`columnEnd`) is past the line width (`columnLimit`) but there's more text to come (`!text.empty()`).

I don't think that completely solves it. You can still have a situation where you write to the last column without more text to come, and then later write additional text that produces a wrap. That wrap is triggered by the `delayedEOLWrap` check in `_WriteToBuffer`, so it doesn't pass through this code, and thus the wrap flag would never be set.

I _think_ the correct place to be handling this, at least for VT code, is in the `_WriteToBuffer` implementation. The legacy console code path may be more complicated - I'm guessing that's going to be somewhere in `WriteCharsLegacy`. There could be all sorts of weird edge cases though. The more I think about it, the more I think it should be left for a follow up PR.

File: src/terminal/adapter/adaptDispatch.cpp [hunk 11]
```diff
@@ -80,11 +80,11 @@ void AdaptDispatch::_WriteToBuffer(const std::wstring_view string)
     // Turn off the cursor until we're done, so it isn't refreshed unnecessarily.
     cursor.SetIsOn(false);
 
-    // The width at which we wrap is determined by the line rendition attribute.
-    auto lineWidth = textBuffer.GetLineWidth(cursorPosition.y);
+    RowWriteState state{
+        .text = string,
+    };
```
Comment (j4james) [inline src/terminal/adapter/adaptDispatch.cpp:85]: I'm afraid this last change is not going to work for me. Because the `AdaptDispatch` class needs to be able to adjust that line width based on the horizontal margins. Storing it in the state should still be OK I think, as long as the initialization of `columnLimit` can be done somewhere here rather than in `TextBuffer::WriteLine`.

Comment (lhecker) [reply src/terminal/adapter/adaptDispatch.cpp:85]: Hmm I see... And I believe I have to modify
```cpp
r.SetWrapForced(wrapAtEOL && state.columnEnd >= lineWidth);
```
to
```cpp
if (wrapAtEOL) {
    r.SetWrapForced(state.columnEnd >= lineWidth);
}
```
because you're going to set `wrapAtEOL` to `false` when writing with margins, right?

Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:85]: I think it probably only needs to be initialized in the one place though: just before `Mode::InsertReplace` check, so the `delayedEOLWrap` code would have had its chance to move to a new line if needed.

Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:85]: > because you're going to set `wrapAtEOL` to `false` when writing with margins, right?

I'm not actually doing anything with the `SetWrapForced` state in my current margin implementation, but I think the best solution is probably to disable all wrapping when horizontal margin mode is enabled. It's a bit like the alt buffer mode in that sense.

File: src/buffer/out/Row.cpp [hunk 12]
```diff
@@ -65,6 +68,65 @@ constexpr OutIt copy_n_small(InIt first, Diff count, OutIt dest)
...
+    _charOffsets{ charOffsets },
+    _beg{ offset },
+    _end{ offset }
+{
+    operator++();
+}
+
+bool RowTextIterator::operator==(const RowTextIterator& other) const noexcept
+{
+    return _beg == other._beg;
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/buffer/out/Row.cpp:82]: I've removed the struct for now, until we need it.

File: src/buffer/out/textBuffer.cpp [hunk 13]
```diff
@@ -376,6 +376,25 @@ bool TextBuffer::_PrepareForDoubleByteSequence(const DbcsAttribute dbcsAttribute
     return fSuccess;
 }
 
+void TextBuffer::ConsumeGrapheme(std::wstring_view& chars) noexcept
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/buffer/out/textBuffer.cpp:379]: In the future I'll replace this function with some ICU code that actually advances the string by an entire grapheme cluster. So, if you have something like å ("a˚") it'll advance by 2 code points and not just 1 like it does now.

It's possible just put the ICU code into `til` like we now and use it throughout the code directly, just like we do it right now with the UTF-16 helpers. But from now on, I'd like to avoid doing that, even if it means writing such `static` methods, because I'd like to keep everything string handling related as close and tight as possible in the future. I think `OutputCellIterator` had the same intention originally and was well meant, but it suffers from being a leaky abstraction. Lots of code is now built around an implicit assumption that `OutputCellIterator` will always advance by exactly 1 or 2 columns. Using the `til` UTF-16 helpers directly elsewhere in our code would have a similar effect in my opinion, because it would equally leak (and potentially incorrectly leak) any assumptions about how `TextBuffer` handles incoming text under normal circumstances.

**Review by DHowett — APPROVED** [review]
